### PR TITLE
fix annotations font color

### PR DESCRIPTION
**Pull request by bensdm**
Correct the font annotation color when using zmin and zmax or zmid https://github.com/plotly/plotly.py/issues/2187
image
image
image

### CODE REVIEW

**Comment by nicolaskruchten** [comment]
Thanks for this pull request! I'll try to review and merge it soon. The build failure is odd and I've seen it before but it doesn't make a ton of sense. I'll look into it :)

File: packages/python/plotly/plotly/figure_factory/_annotated_heatmap.py [hunk 1]
```diff
@@ -264,34 +262,48 @@ def get_text_color(self):
...
         """
         min_text_color, max_text_color = _AnnotatedHeatmap.get_text_color(self)
-        z_mid = _AnnotatedHeatmap.get_z_mid(self)
+        if zmin is None or zmax is None:
+            zmin, zmax = _AnnotatedHeatmap.get_z_min_max(self)
+        if zmid is None:
+            zmid = (zmax + zmin) / 2
+        if min_text_color == max_text_color:
+            # diverging colorscale
+            mid_text_color = "#000000"
```
Comment (nicolaskruchten) [inline packages/python/plotly/plotly/figure_factory/_annotated_heatmap.py:293]: I'm not sure this is a great default... this is assuming the diverging scale is light in the middle?

File: packages/python/plotly/plotly/figure_factory/_annotated_heatmap.py [hunk 2]
```diff
@@ -264,34 +262,48 @@ def get_text_color(self):
...
         """
         Get annotations for each cell of the heatmap with graph_objs.Annotation
 
         :rtype (list[dict]) annotations: list of annotations for each cell of
             the heatmap
         """
         min_text_color, max_text_color = _AnnotatedHeatmap.get_text_color(self)
-        z_mid = _AnnotatedHeatmap.get_z_mid(self)
+        if zmin is None or zmax is None:
+            zmin, zmax = _AnnotatedHeatmap.get_z_min_max(self)
```
Comment (nicolaskruchten) [inline packages/python/plotly/plotly/figure_factory/_annotated_heatmap.py:288]: here it seems like we overwrite both if either is missing? probably best not to do that I think.

File: packages/python/plotly/plotly/figure_factory/_annotated_heatmap.py (hunk 1 above)
Comment (nicolaskruchten) [reply packages/python/plotly/plotly/figure_factory/_annotated_heatmap.py:293]: I wonder if we shouldn't accept a third color in `font_colors` for this and only set this one here if it's not present?

**Comment by nicolaskruchten** [comment]
Thanks for this PR! It's prompted me to make this one https://github.com/plotly/plotly.py/pull/2921 which is a bit more targeted and doesn't change any method signatures etc. I'd love your thoughts on the other one, and if we merge that then you can try to build on it to address the color problem in diverging colorscales?

File: packages/python/plotly/plotly/figure_factory/_annotated_heatmap.py (hunk 2 above)
Comment (bensdm) [reply packages/python/plotly/plotly/figure_factory/_annotated_heatmap.py:288]: From what I understand from the doc https://plotly.github.io/plotly.py-docs/generated/plotly.graph_objects.Heatmap.html "zmax – Sets the upper bound of the color domain. Value should have the same units as in z and if set, zmin must be set as well." we can not pass only one of these 2 thats why i overwrite both

File: packages/python/plotly/plotly/figure_factory/_annotated_heatmap.py (hunk 1 above)
Comment (bensdm) [reply packages/python/plotly/plotly/figure_factory/_annotated_heatmap.py:293]: Yes can be an option but is there any diverging colorscale where the it's not the case?

Comment (nicolaskruchten) [reply packages/python/plotly/plotly/figure_factory/_annotated_heatmap.py:293]: Well, users can pass in their own colorscale, they're not limited to just the built-in ones.

File: packages/python/plotly/plotly/figure_factory/_annotated_heatmap.py (hunk 2 above)
Comment (nicolaskruchten) [reply packages/python/plotly/plotly/figure_factory/_annotated_heatmap.py:288]: This is true of the underlying heatmap trace, yes, but this figure-factory already does its own bounds-computation, so we may as well reuse it. I've already addressed this comment in #2921.

**Pull request closed without merging**